PresentData/PresentMonTraceConsumer.cpp

```diff
@@ -644,20 +680,27 @@ void PMTraceConsumer::HandleDXGKEvent(EVENT_RECORD* pEventRecord)
             eventIter->second->Hwnd = mMetadata.GetEventData<uint64_t>(pEventRecord, L"hWindow");
         }
 
-        if (eventIter->second->PresentMode == PresentMode::Hardware_Legacy_Copy_To_Front_Buffer &&
-            eventIter->second->ScreenTime != 0) {
-            // This is a fullscreen or DWM-off blit where all work associated was already done, so it's on-screen
-            // It was deferred to here because there was no way to be sure it was really fullscreen until now
-            CompletePresent(eventIter->second);
-        }
+        // Create a temporary copy of the shared_ptr since we may erase the iterator before we are done with its data.
+        std::shared_ptr<PresentEvent> event = eventIter->second;
 
-        if (eventIter->second->ThreadId != hdr.ThreadId) {
-            if (eventIter->second->TimeTaken == 0) {
-                eventIter->second->TimeTaken = hdr.TimeStamp.QuadPart - eventIter->second->QpcTime;
+        if (event->ThreadId != hdr.ThreadId) {
+            // In this case the same presentEvent may have two entries in mPresentByThreadId. Remove the extra one.

```

Comment:
Are you sure about this? I don't think that's true.

Reply:
Not sure about this one, that was my inference from reading this case block, but I could well have misread.

I think the thing that trips me up here is that we look up the event in mPresentByThreadId by hdr.ThreadId , but expect some cases where event->ThreadId.

This should only be possible if the ThreadId of a present in mPresentByThreadId is modified without changing the key, or if a present has its key in mPresentByThreadId modified without changing its ThreadId. Is there somewhere in the code where we do this?

Reply:
Oh, it can move to a different thread ID in the map, but this isn't the right place to put logic related to that.

A single present can be inserted into the map with two different thread IDs if:
1. `CreatePresent` is called directly by a D3D9/DXGI event. The thread ID stored in the present is this one, which is the thread the app called on.
2. If the present was batched, then `FindOrCreatePresent` will also add it to the map, keyed by the thread ID that the driver called into the kernel. This isn't currently stored in the present structure.

The `RuntimePresentStop` logic will remove that first entry, and the second one will be removed once we don't expect anything that's correlated by thread for the present anymore. We probably want to add a second thread ID to the present struct, which is assigned by `FindOrCreatePresent` when it gets added to the map a second time, so that can be cleaned up generically.

Reply:
Ok, in the latest iteration I replaced WasBatched with the more informative DriverBatchThreadId (0 means not batched), and now remove based on this Id too.

I also noticed that FindOrCreatePresent removes from the mPresentsByProcess list, which was previously assumed to always contain all presents. I removed the assert that something has to be removed in RemoveLostPresent.

Reply:
Looks good.